File: src/main/java/xyz/upperlevel/hgame/event/EventListener.java
package xyz.upperlevel.hgame.event;

import lombok.Data;
import lombok.EqualsAndHashCode;

import java.util.function.Consumer;

@EqualsAndHashCode
@Data
public abstract class EventListener<E extends Event> {
    private final Class<?> clazz;
    private final byte priority;

    public EventListener(Class<?> clazz, byte priority) {
        this.clazz = clazz;
        this.priority = priority;
    }

    public EventListener(Class<?> clazz) {
        this(clazz, EventPriority.NORMAL);
    }

    public abstract void call(E event);

    public static <E extends Event> EventListener<E> listener(Class<E> clazz, Consumer<E> consumer, byte priority) {
        return new SimpleEventListener<>(clazz, priority, consumer);
    }

    public static <E extends Event> EventListener<E> listener(Class<E> clazz, Consumer<E> consumer) {
        return new SimpleEventListener<>(clazz, EventPriority.NORMAL, consumer);
    }

    public static <E extends Event> EventListener<E> of(Class<E> clazz, Consumer<E> consumer, byte priority) {
        return new SimpleEventListener<>(clazz, priority, consumer);
    }

    public static <E extends Event> EventListener<E> of(Class<E> clazz, Consumer<E> consumer) {
        return new SimpleEventListener<>(clazz, EventPriority.NORMAL, consumer);
    }
}

File: src/main/java/xyz/upperlevel/hgame/event/EventListener.kt
package xyz.upperlevel.hgame.event

abstract class EventListener<E : Event> (
        val clazz: Class<*>,
        val priority: Byte = EventPriority.NORMAL) {
    abstract fun call(event: E)

    override fun equals(other: Any?): Boolean {
        if (other === this) return true
        if (other == null) return false
        if (other.javaClass != this.javaClass) return false
        other as EventListener<*> // Come on kotlin
        if (this.clazz != other.clazz) return false
        return this.priority == other.priority
    }

    override fun hashCode(): Int {
        val prime = 59
        var result = 1
        result = result * prime + this.clazz.hashCode()
        result = result * prime + this.priority
        return result
    }

    companion object {
        fun <E : Event> listener(clazz: Class<E>, consumer: (E) -> Unit, priority: Byte = EventPriority.NORMAL): EventListener<E> {
            return SimpleEventListener(clazz, consumer, priority)
        }

        fun <E : Event> of(clazz: Class<E>, consumer: (E) -> Unit, priority: Byte = EventPriority.NORMAL): EventListener<E> {
            return SimpleEventListener(clazz, consumer, priority)
        }
    }
}
